lib/iris/analysis/__init__.py

```diff
@@ -2338,180 +2347,117 @@ def group(self):
-        # tuple group indices, and a list of the respective bounds of those
-        # indices.
-        for key_slice in self._slices_by_key.values():
-            if isinstance(key_slice, tuple):
-                indices = key_slice
-            else:
-                indices = tuple(range(*key_slice.indices(self._stop)))
-
-            groupby_indices.append(indices)
-            groupby_bounds.append((indices[0], indices[-1]))
-
-        # Create new shared bounded coordinates.
         for coord, dim in self._shared_coords:
             climatological_coord = (
                 self.climatological and coord.units.is_time_reference()
             )
             if coord.points.dtype.kind in "SU":
                 if coord.bounds is None:
-                    new_points = []
+                    new_points_list = []
                     new_bounds = None
                     # np.apply_along_axis does not work with str.join, so we
                     # need to loop through the array directly. First move axis
                     # of interest to trailing dim and flatten the others.
                     work_arr = np.moveaxis(coord.points, dim, -1)
                     shape = work_arr.shape
                     work_shape = (-1, shape[-1])
-                    new_shape = (len(self),)
+                    new_shape: tuple[int, ...] = (len(self),)
                     if coord.ndim > 1:
                         new_shape += shape[:-1]
                     work_arr = work_arr.reshape(work_shape)
 
-                    for indices in groupby_indices:
+                    for indices in self._groupby_indices:
                         for arr in work_arr:
-                            new_points.append("|".join(arr.take(indices)))
+                            new_points_list.append("|".join(arr.take(indices)))
 
                     # Reinstate flattened dimensions. Aggregated dim now leads.
-                    new_points = np.array(new_points).reshape(new_shape)
+                    new_points = np.array(new_points_list).reshape(new_shape)
 
                     # Move aggregated dimension back to position it started in.
                     new_points = np.moveaxis(new_points, 0, dim)
                 else:
                     msg = (
-                        "collapsing the bounded string coordinate {0!r}"
-                        " is not supported".format(coord.name())
+                        "collapsing the bounded string coordinate"
+                        f" {coord.name()!r} is not supported"
                     )
                     raise ValueError(msg)
             else:
-                new_bounds = []
+                new_bounds_list = []
                 if coord.has_bounds():
                     # Derive new coord's bounds from bounds.
                     item = coord.bounds
-                    maxmin_axis = (dim, -1)
+                    maxmin_axis: Union[int, tuple[int, int]] = (dim, -1)
```

Comment:
Yes, in most cases `mypy` can automatically infer the type, but in this case, it gets it wrong because the implementation actually allows two possible types.